File: src/netplot/client/NetplotClient.java
/*****************************************************************************************
 *                             Copyright 2009 Paul Austen                                *
 *                                                                                       *
 * This program is distributed under the terms of the GNU Lesser General Public License  *
 *****************************************************************************************/

package netplot.client;

import java.net.*;
import java.util.Calendar;
import java.util.StringTokenizer;
import java.io.*;
import netplot.KeyWords;
import netplot.LineProcessor;

import java.util.*;
import org.jfree.data.time.Millisecond;

/**
 * Responsible for providing a client interface to the netplot server.
 */
public class NetplotClient
{
  public boolean    debug;
  public double     serverVersion;
  Socket            socket;
  BufferedReader    br;
  PrintWriter       pw;
  int               serverResponseTimeout=30000;
  boolean           cacheEnabled; //In cached plot mode the commands to add plot points to a graph or graphs 
                                  //are cached until the update() method is called. This means that all plot
                                  //points are sent to the server together. When not in fast mode (default), 
                                  //each time addPlotValues or addXYPlotValues is called then data is sent 
                                  //to the server to add these plot points.
  Vector<String>    plotValueCache;  
  
  public NetplotClient()
  {
  }
  public NetplotClient(boolean debug)
  {
    this.debug=debug;
    plotValueCache = new Vector<String>();
  }

  /**
   * Display debug messages if required
   */
  void debugPrint(String message)
  {
    if( debug)
      Println("DEBUG: "+message);
  }
  
  /**
   * display a message.
   * 
   * @param message
   */
  public static void Println(String message)
  {
    System.out.println(message);
  }
  
  /**
   * Connect to the server running the netplot GUI (Java) program
   * @param hostAddress The netplot server address.
   * @param port        The netplot server TCP/IP port.
   */
  public void connect(String hostAddress, int port) throws NetplotClientException, IOException
  {
    debugPrint("Connecting to "+hostAddress+":"+port);
    socket = new Socket(hostAddress,port);
    //Set the read timeout on the server response
    socket.setSoTimeout(serverResponseTimeout);
    br = new BufferedReader( new InputStreamReader(socket.getInputStream()));
    pw = new PrintWriter(socket.getOutputStream(), true);
    debugPrint("Connected to "+hostAddress+":"+port);
    //Wait for initial connection message
    String line = br.readLine();
    debugPrint("line="+line);
    boolean netPlotServer=true;
    StringTokenizer strTok = new StringTokenizer(line,"=");
    if( line.indexOf("netplot_version=") != 0 || strTok.countTokens() != 2 )
    {
      netPlotServer=false;
    }
    if(netPlotServer) 
    {
      try
      {
        strTok.nextToken();
        serverVersion=Double.parseDouble(strTok.nextToken());
      }
      catch(Exception e) 
      {
        netPlotServer=false;
      }
    }
    if( !netPlotServer )
    {
      throw new NetplotClientException(hostAddress+":"+port+" is not a netplot server. Received "+line);
    }
  }
  
  /**
   * Disconnect form the netplot server and clean up resources.
   */
  public void disconnect()
  {
    if( socket != null )
    {
      try
      {
        socket.close();
      }
      catch(IOException e) {}
      socket=null;
      br=null;
      pw.close();
      pw=null;
    }
  }
  
  /**
   * Get the netplot server version received when we connected to the netplot server.
   * 
   * @return The server version
   */
  public double getServerVersion()
  {
    return serverVersion;
  } 

  
  /**
   * Send a command to the netplot server
   * 
   * @param cmd the command
   */
  void sendCmd(String cmd) throws NetplotClientException, IOException
  {
    debugPrint("CMD: "+cmd);
    pw.println(cmd);
    //Get response from server
    //We expect and OK response from the server
    String line = br.readLine();
    debugPrint(line);
    if( line.indexOf("ERROR: ") == 0)
    {
      throw new NetplotClientException(line);
    }
  }
  
  /**
   * Send a set command
   * 
   * @param var     The variable name
   * @param value   The variable value
   * @throws NetplotClientException
   */
  void sendSetCmd(String var, String value) throws NetplotClientException, IOException
  {
    sendCmd("set "+var+"="+value);
  }

  /**
   * Set the plot grid GUI layout
   */
  public void setGrid(int rows, int columns) throws NetplotClientException, IOException
  {
    sendSetCmd(KeyWords.GRID,""+rows+","+columns);
  }
  
  /**
   * Set the GUI window title
   * 
   * @param windowTitle The title text
   */
  public void setWindowTitle(String windowTitle) throws NetplotClientException, IOException
  {
    sendSetCmd(KeyWords.FRAME_TITLE, windowTitle);
  }
  
  /**
   * Enable/disable the chart legend
   * 
   * @param enabled If true the legend is enabled.
   */
  public void setChartLegendEnabled(boolean enabled) throws NetplotClientException, IOException
  {
    sendSetCmd(KeyWords.ENABLE_LEGEND,""+enabled);
  }

  /**
   * Init the chart
   */
  public void init() throws NetplotClientException, IOException
  {
    sendCmd(KeyWords.INIT);
  }

  /**
   * Check for a valid plot type.
   * 
   * @param plotType The plot type string 
   * @return true if valid
   */
  public boolean isValidPlotType(String plotType)
  {
    if( plotType.equals("time") ||
        plotType.equals("xy")   ||
        plotType.equals("bar")  ||
        plotType.equals("dial") ) 
    {
      return true;
    }
    return false;  
  }
  
  public void setPlotType(String plotType) throws NetplotClientException, IOException
  {
    if( !isValidPlotType(plotType) )
    {
      throw new NetplotClientException(plotType+" is an invalid plot type");
    }
    sendSetCmd(KeyWords.GRAPH,plotType);
  }
  
  public void setPlotType(String plotType, String title) throws NetplotClientException, IOException
  {
    setPlotType(plotType);
    //If we have a 
    if( title != null && title.length() > 0 )
    {
      sendSetCmd(KeyWords.PLOT_TITLE,title);
    }
    init();
  }
    
  /**
   * Add a plot trace
   * 
   * @param plotConfig The config for the plot
   */
  public void addPlot(PlotConfig plotConfig) throws NetplotClientException, IOException
  {
    if( plotConfig != null)
    {
      sendSetCmd(KeyWords.PLOT_NAME,plotConfig.plotName);
      sendSetCmd(KeyWords.X_AXIS_NAME,plotConfig.xAxisName);
      sendSetCmd(KeyWords.Y_AXIS_NAME,plotConfig.yAxisName);
      sendSetCmd(KeyWords.ENABLE_LINES,""+plotConfig.enableLines);
      sendSetCmd(KeyWords.LINE_WIDTH,""+plotConfig.lineWidth);
      sendSetCmd(KeyWords.ENABLE_SHAPES,""+plotConfig.enableShapes);
      sendSetCmd(KeyWords.ENABLE_AUTOSCALE,""+plotConfig.enableAutoScale);
      sendSetCmd(KeyWords.MIN_SCALE_VALUE,""+plotConfig.minScaleValue);
      sendSetCmd(KeyWords.MAX_SCALE_VALUE,""+plotConfig.maxScaleValue);
      sendSetCmd(KeyWords.MAX_AGE_SECONDS,""+plotConfig.maxAgeSeconds);
      sendSetCmd(KeyWords.ENABLE_LOG_Y_AXIS,""+plotConfig.enableLogYAxis);
      sendSetCmd(KeyWords.ENABLE_ZERO_ON_X_SCALE,""+plotConfig.enableZeroOnXAxis);
      sendSetCmd(KeyWords.ENABLE_ZERO_ON_Y_SCALE,""+plotConfig.enableZeroOnYAxis);
      sendSetCmd(KeyWords.TICK_COUNT,""+plotConfig.tickCount);
    }
    sendCmd(KeyWords.ADD_PLOT);
  }
  
  /**
   * Add values to plot/s
   * 
   * @param values An array of values to be plotted, one element for each 
   *               plot added. The number of plots is determined by the number 
   *               of times addPlot has been called. Therefore the first element 
   *               in the list is added to the first plot, second to the seconds 
   *               and so on.
   */
  public void addPlotValues(double values[]) throws NetplotClientException, IOException
  {
    if( values.length == 0 )
    {
      return;
    }
    StringBuffer strBuf = new StringBuffer(""+values[0]);
    boolean firstValue=true;
    for( double value : values )
    {
      if( firstValue )
      {
        firstValue=false;
        continue;
      }
      strBuf.append(","+value);
    }
    debugPrint("Adding plot values: "+strBuf);
    if( cacheEnabled ) {
      plotValueCache.add(strBuf.toString());
    }
    else {
      sendCmd(strBuf.toString());
    }
  }
  
  /**
   * Add values to an XY plot
   * @param plotIndex  The index of the plot (0 = first plot added, 1=second etc)
   * @param xValue     The value on the X axis
   * @param yValue     The value on the Y axis
   */
  public void addXYPlotValues(int plotIndex, double xValue, double yValue) throws NetplotClientException, IOException
  {
    debugPrint("Adding XY plot "+plotIndex+" values: "+xValue+","+yValue);
    if( cacheEnabled ) {
      plotValueCache.add(plotIndex+":"+xValue+":"+yValue);
    }
    else {
      sendCmd(plotIndex+":"+xValue+":"+yValue);
    }
  }
 
  /**
   * Add values to an XY plot
   * @param plotIndex  The index of the plot (0 = first plot added, 1=second etc)
   * @param xValue     The value on the X axis
   * @param yValue     The value on the Y axis
   */
  public void addTimeSeriesPlotValue(int plotIndex, Millisecond ms, double yValue) throws NetplotClientException, IOException
  {
    debugPrint("Adding TimeSeries plot "+plotIndex+" Millisecond "+ms+" value "+yValue);
    Calendar calendar = Calendar.getInstance();
    calendar.setTime( ms.getStart() );
    String dateString = this.getDateString(calendar, ms.getMillisecond() );
    sendCmd(plotIndex+":"+dateString+":"+yValue);
  }
  
  /**
   * Get date string.
   * @param date The Date object to be converted to a string.
   * @param ms The MilliSecond value to be included in the Date String.
   * @return The String representation of the Date.
   * 
   * The format of the string returned is 
   * <YEAR>;<MONTH>;<DAY OF MONTH>;<HOUR OF DAY>;<MINUTE>;<SECOND>;<MILLISECOND>
   * 
   * <YEAR>  = Year (1900-9999)
   * <MONTH> = Month (1-12)
   * <DAY OF MONTH> = Month (1-31)
   * <HOUR OF DAY>  = Hour of day (0-23)
   * <MINUTE>       = Minute (0-59)
   * <SECOND>       = Second (0-59)
   * <MILLISECOND>  = Millisecond (0-999)
   * E.G
   * 2013;00;02;23;10;05;587
   * 
   * Represents
   * 
   * 23:20:05:587 on 1 Jan 2013
   * 
   */
  private String getDateString(Calendar calendar, long ms) {
    StringBuffer sb= new StringBuffer();

    sb.append(calendar.get(Calendar.YEAR));
    sb.append(LineProcessor.TIMESTAMP_DELIM);
    sb.append(calendar.get(Calendar.MONTH)+1);
    sb.append(LineProcessor.TIMESTAMP_DELIM);
    sb.append(calendar.get(Calendar.DAY_OF_MONTH));
    sb.append(LineProcessor.TIMESTAMP_DELIM);
    sb.append(calendar.get(Calendar.HOUR_OF_DAY));
    sb.append(LineProcessor.TIMESTAMP_DELIM);
    sb.append(calendar.get(Calendar.MINUTE));
    sb.append(LineProcessor.TIMESTAMP_DELIM);
    sb.append(calendar.get(Calendar.SECOND));
    sb.append(LineProcessor.TIMESTAMP_DELIM);
    sb.append(ms);

    return sb.toString();
  }

  /**
   * Clear the selected plot
   */
  public void clear(int plotIndex) throws NetplotClientException, IOException
  {
    sendCmd(KeyWords.CLEAR+" "+plotIndex);
  }

  /**
   * Clear the selected plot
   */
  public void replot(int plotIndex) throws NetplotClientException, IOException
  {
    sendCmd(KeyWords.REPLOT+" "+plotIndex);
  }

  /**
   * Enable/Disable the GUI status messages.
   * If CPU bouind then not prointing the status messages to the status history may speed up plotting.
   */
  public void enableStatusMessages(boolean enabled) throws NetplotClientException, IOException
  {
    if( enabled ) {
      sendCmd("enable_status 1");
    }
    else {
      sendCmd("enable_status 0");      
    } 
  }
  
  /**
   * Enable/Disable the cache.
   * If enabled then the update method must be called to plot the data.
   * 
   * @param enabled
   */
  public void enableCache(boolean enabled) {
    this.cacheEnabled=enabled;
  }

  /**
   * Send all plotValueCache plot points. Only call this when cacheEnabled is True
   */
  public void update() throws NetplotClientException, IOException {
    int cmdCount = plotValueCache.size();
    
    //If there is nothing to send
    if( cmdCount == 0 ) {
      //quit
      return;
    }
      
    //Build a single string containing all the commands
    StringBuffer cmd = new StringBuffer();
    for( String plotValue : plotValueCache ) {
      cmd.append(plotValue+"\n");
    }
    //Empty the cache
    plotValueCache.removeAllElements();
      
    debugPrint("CMD: "+cmd);
    pw.println(cmd);
    
    //Wait for responses
    int rxCmdCount=0;
    while( rxCmdCount < cmdCount ) {
      //Get response from server
      //We expect and OK response from the server
      String line = br.readLine();
      debugPrint(line);
      if( line.indexOf("ERROR: ") == 0)
      {
        throw new NetplotClientException(line);
      }
      rxCmdCount++;
    }
  }
      

  
}

File: python_client/netplot_client.py
#!/usr/bin/python

#/*****************************************************************************************
#*                             Copyright 2009 Paul Austen                                *
#*                                                                                       *
#* This program is distributed under the terms of the GNU Lesser General Public License  *
#*****************************************************************************************/

import socket
import types
import types

VALID_PLOT_TYPES=['time','bar','xy', 'dial']

PLOT_TITLE		= "plot_title"
PLOT_NAME		= "plot_name"
X_AXIS_NAME		= "x_axis_name"
Y_AXIS_NAME		= "y_axis_name"
ENABLE_LINES		= "enable_lines"
ENABLE_SHAPES		= "enable_shapes"
ENABLE_AUTOSCALE	= "enable_autoscale"
MIN_SCALE_VALUE		= "min_scale_value"
MAX_SCALE_VALUE		= "max_scale_value"
MAX_AGE_SECONDS		= "max_age_seconds"
ENABLE_LOG_Y_AXIS	= "enable_log_y_axis"
ENABLE_ZERO_ON_X_SCALE  = "enable_zero_on_x_scale"
ENABLE_ZERO_ON_Y_SCALE  = "enable_zero_on_y_scale"
FRAME_TITLE		= "frame_title"
ENABLE_LEGEND		= "enable_legend"
TICK_COUNT 		= "tick_count"

class NetPlotError(Exception):
  pass
  
class PlotConfig:
  def __init__(self):
    self.plotName=""
    self.xAxisName=""
    self.yAxisName=""
    self.enableLines=1
    self.enableShapes=1
    self.enableAutoScale=1
    self.minScaleValue=0
    self.maxScaleValue=1E6
    self.maxAgeSeconds=3600
    self.enableLogYAxis=0
    self.enableZeroOnXAxis=1
    self.enableZeroOnYAxis=1
    self.tickCount=0
    
class NetPlot:
    
  TIMESTAMP_DELIM=";"

  def __init__(self,debug=0):
    #If debug set on then debug messages will be shown on stdout
    self.__debug=debug
    self.__hostAddress=None
    self.__port=None
    self.__serverVersion=None
    self.__cacheEnabled=False   #In cached plot mode the commands to add plot points to a graph or graphs 
                                #are cached until the update() method is called. This means that all plot
                                #points are sent to the server together. When not in fast mode (default), 
                                #each time addPlotValues or addXYPlotValues is called then data is sent 
                                #to the server to add these plot points.
    self.__plotValueCache=[]    
    self.sock = None


  def __debugPrint(self, message):
    """Display debug messages if required"""
    if self.__debug:
      print "DEBUG: %s" % (message)

  def connect(self, hostAddress, port):
    """Connect to the server running the netplot GUI (Java) program"""
    self.__hostAddress=hostAddress
    self.__port=port
    try:
        self.__debugPrint('Connecting to %s:%d' % (self.__hostAddress, self.__port) )
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((self.__hostAddress, self.__port))
    except socket.error:
        raise socket.error("netplot connect failed to %s:%d" % (self.__hostAddress, self.__port) )
        
    self.__debugPrint('Connected to %s:%d' % (self.__hostAddress, self.__port) )
    #Wait for initial connection message
    rxData = self.sock.recv(256)
    self.__debugPrint("rxData=%s" % (rxData))
    netPlotServer=1
    elems=rxData.split('=')
    if rxData.find('netplot_version=') != 0 or len(elems) != 2:
      netPlotServer=0
    if netPlotServer:
      try:
        self.__serverVersion=float(elems[1])
      except ValueError:
        netPlotServer=0
    if netPlotServer == 0:
      raise NetPlotError("%s:%d is not a netplot server. Received %s" % (self.__hostAddress, self.__port, rxData) )
    self.sock.setblocking(0)
    
  def disconnect(self):
    """Close a connection to a netplot server"""
    if self.sock != None:
      self.sock.close()
      
  def getServerVersion(self):
    return self.__serverVersion
    
  def sendCmd(self, cmd):
    """Send a command to the netplot server"""
    self.__debugPrint('CMD: %s' % (cmd) )
    self.sock.send('%s\n' % (cmd) )
    #Wait for response
    while 1:
	    try:
	    	rxData = self.sock.recv(256)
	    	self.__debugPrint("rxData=%s" % (rxData))
	    	if rxData.find('OK') == 0:
	    		break
	    	elif rxData.find('ERROR: ') == 0:
	    	  raise NetPlotError(rxData)
	    except socket.error:
	      pass
    
  def setGrid(self, rows, columns):
    """Set the number of graphs and their layout"""
    self.sendCmd("set grid=%d,%d" % (rows, columns) )

  def setWindowTitle(self, windowTitle):
    """Set the frame title"""
    self.sendCmd("set %s=%s" % (FRAME_TITLE,windowTitle) )
    
  def setChartLegendEnabled(self, enabled):
    """enable/disable legends on the chart"""
    varValue="false"
    if enabled:
      varValue="true"
    self.sendCmd("set %s=%s" % (ENABLE_LEGEND,varValue) )
    
  def init(self):
    """Initialise the plot. May be called to clear a plot"""
    self.sendCmd("init")

  def setPlotType(self, plotType, title=None):
    if plotType not in VALID_PLOT_TYPES:
      raise NetPlotError("%s is an invalid plot type" % (plotType) )
    self.sendCmd("set graph=%s" % (plotType) )
    if title != None:
      self.sendCmd("set %s=%s" % (PLOT_TITLE,title) )
    self.init()
    #By default we enable status messages
    self.enableStatusMessages(True)
    
  def addPlot(self, plotConfig=None):
    if plotConfig != None:
      if plotConfig.plotName != None:
      	self.sendCmd("set %s=%s" % (PLOT_NAME,plotConfig.plotName) )
      if plotConfig.xAxisName != None:
        self.sendCmd("set %s=%s" % (X_AXIS_NAME,plotConfig.xAxisName) )
      if plotConfig.yAxisName != None:
        self.sendCmd("set %s=%s" % (Y_AXIS_NAME,plotConfig.yAxisName) )
      if plotConfig.enableLines != None:
        varValue="false"
        if plotConfig.enableLines:
          varValue="true"
        self.sendCmd("set %s=%s" % (ENABLE_LINES,varValue) )
      if plotConfig.enableShapes != None:
        varValue="false"
        if plotConfig.enableShapes:
          varValue="true"
        self.sendCmd("set %s=%s" % (ENABLE_SHAPES,varValue) )
      if plotConfig.enableAutoScale != None:
        varValue="false"
        if plotConfig.enableAutoScale:
          varValue="true"
        self.sendCmd("set %s=%s" % (ENABLE_AUTOSCALE,varValue) )
      if plotConfig.minScaleValue != None:
        self.sendCmd("set %s=%s" % (MIN_SCALE_VALUE,str(plotConfig.minScaleValue)) )
      if plotConfig.maxScaleValue != None:
        self.sendCmd("set %s=%s" % (MAX_SCALE_VALUE,str(plotConfig.maxScaleValue)) )
      if plotConfig.maxAgeSeconds != None:
        self.sendCmd("set %s=%s" % (MAX_AGE_SECONDS,str(plotConfig.maxAgeSeconds)) )
      if plotConfig.enableLogYAxis != None:
        varValue="false"
        if plotConfig.enableLogYAxis:
          varValue="true"
        self.sendCmd("set %s=%s" % (ENABLE_LOG_Y_AXIS,varValue) )
      if plotConfig.enableZeroOnXAxis != None:
        varValue="false"
        if plotConfig.enableZeroOnXAxis:
          varValue="true"
        self.sendCmd("set %s=%s" % (ENABLE_ZERO_ON_X_SCALE,varValue) )
      if plotConfig.enableZeroOnYAxis != None:
        varValue="false"
        if plotConfig.enableZeroOnYAxis:
          varValue="true"
        self.sendCmd("set %s=%s" % (ENABLE_ZERO_ON_Y_SCALE,varValue) )
      if plotConfig.tickCount != None:
        self.sendCmd("set %s=%s" % (TICK_COUNT,str(plotConfig.tickCount)) )  	       		
    self.sendCmd("add_plot")

  def __getValue(self, value):
    if type(value) == types.StringType:
      return value
    elif type(value) == types.LongType:
      return "%d" % (value)
    elif type(value) == types.IntType:
      return "%d" % (value)
    elif type(value) == types.FloatType:
      return "%f" % (value)
    else:
      self.__debugPrint("%s is of an unsupported type (%s), cannot plot" % (repr(value), repr(type(value)))) 

  def _getDateTimeString(self, dateTimeObj):
    """Get the date time string in the required format to plot time series plots."""   
    return ""+str(dateTimeObj.year)+NetPlot.TIMESTAMP_DELIM+\
              str(dateTimeObj.month)+NetPlot.TIMESTAMP_DELIM+\
              str(dateTimeObj.day)+NetPlot.TIMESTAMP_DELIM+\
              str(dateTimeObj.hour)+NetPlot.TIMESTAMP_DELIM+\
              str(dateTimeObj.minute)+NetPlot.TIMESTAMP_DELIM+\
              str(dateTimeObj.second)+NetPlot.TIMESTAMP_DELIM+\
              str(dateTimeObj.microsecond/1000)
    
    
  def addPlotValues(self, values):
    """Add the plot values

       values must be a list of values that contains one element for each 
       plot added. The number of plots is determined by the number of 
       times addPlot has been called. Therefore the first element in 
       the list is added to the first plot, second to the seconds and so on. 
    """ 
    self.__debugPrint('Adding plot values: %s' % (repr(values))  )
    cmdString=""
    firstValue=1
    for value in values:
        if firstValue:
          cmdString=self.__getValue(value)
        else:
          cmdString="%s,%s" % (cmdString,self.__getValue(value))
        firstValue=0
        if self.__cacheEnabled:
          self.__plotValueCache.append(cmdString)
          self.updateIfRequired()
        else:
          self.sendCmd(cmdString)
        
  def updateIfRequired(self):
      """Call a plot update if we need to do so because there are a lot of plot points outstanding."""
      if self.__cacheEnabled:
          #Ensure we don't have more than 50 plot points outstanding
          if len(self.__plotValueCache) > 50:
              self.update()

  def addTimePlotValue(self, plotIndex, dateTimeObj, plotValue):
    """The recommended way to send time series plot values where the time is passed from the 
       netplot client to the netplot server (GUI)."""
    timeStampStr = self._getDateTimeString(dateTimeObj)
    cmdString="%d:%s:%s" % (plotIndex,timeStampStr, self.__getValue(plotValue))
    if self.__cacheEnabled:
      self.__plotValueCache.append(cmdString)
      self.updateIfRequired()
      
    else:
      self.sendCmd(cmdString)
    
  def addXYPlotValues(self, plotIndex, xValue, yValue):
    """Add the XY plot values
    """ 
    self.__debugPrint('Adding XY plot values: %f:%f' % (xValue, yValue)  )
    msg="%d:%E:%E" % (plotIndex,xValue,yValue)
    if self.__cacheEnabled:
      self.__plotValueCache.append(msg)
      self.updateIfRequired()
    else:
      self.sendCmd(msg)

  def clear(self, plotIndex):
    """Clear the plot referenced by plotIndex"""
    self.__debugPrint('Clearing plot %d' % (plotIndex)  )
    self.sendCmd("clear %d" % (plotIndex) )
    
  def replot(self, plotIndex):
    """replot causes subsequent values to overwrite previous ones.
       This is more efficient than clear and stops screen flicker
       when all plot points are erased.
    """
    self.__debugPrint('Clearing plot %d' % (plotIndex)  )
    self.sendCmd("replot %d" % (plotIndex) )

  def enableStatusMessages(self, enabled):
    """Enable/disable the status messages on the display window"""
    if enabled:
      self.sendCmd("enable_status 1")
    else:	
      self.sendCmd("enable_status 0")
      
  def enableCache(self, enabled):
    """Enable/Disable the plot cache"""
    self.__cacheEnabled=enabled
	
  def update(self):
    """Send all plotValueCache plot points.
       Only call this when__cacheEnabled is True"""
    cmdCount=len(self.__plotValueCache)
    #If there is nothing to send
    if cmdCount == 0:
    	#quit
      return
      
    #Build a single string containing all the commands
    cmd=""
    for plotValue in self.__plotValueCache:
      cmd="%s%s\n" % (cmd,plotValue)
    #empty the cache
    self.__plotValueCache = []
      
    self.__debugPrint('CMD: %s' % (cmd) )
    self.sock.send('%s\n' % (cmd) )
    #Wait for responses
    rxCmdCount=0
    while rxCmdCount < cmdCount:
      try:
        rxData = self.sock.recv(65536)
        rxCmdCount = rxCmdCount + self.__processResponse(rxData)
      except socket.error:
        pass
       
  def __processResponse(self, rxData):
    """Process the data received from the netplot server
    	Return: The number of ok responses received
    	Throws NetPlotError if the server returns an error
    """
    if rxData == None:
    	raise NetPlotError("rxData from server == None")
    	
    if len(rxData) == 0:
      raise NetPlotError("No rxData received from netplot server")
    
    offset=0
    #Process all responses received in the rxData
    while offset < len(rxData):
      if rxData[offset:].find('OK') == 0:
        offset=offset+2
        
      elif rxData[offset:].find('ERROR: ') == 0:
	      raise NetPlotError(rxData)
	      
      if rxData[offset] == '\n' or\
         rxData[offset] == '\r':
        offset=offset+1
      
    #Return the number of OK responses received
    return offset/2
     
